Sources/PointFree/Newsletters.swift

```diff
@@ -1,13 +1,43 @@
 import ApplicativeRouter
+import Either
 import HttpPipeline
 import Foundation
 import Prelude
 import Tuple
 
 let expressUnsubscribeMiddleware =
-{ (conn: Conn<StatusLineOpen, T4<Database.User?, Database.User.Id, Database.EmailSetting.Newsletter, Prelude.Unit>>) -> IO<Conn<StatusLineOpen, Prelude.Unit>> in
+  unsubscribeMiddleware
+    >-> redirect(to: .secretHome)
 
-  let (_, userId, newsletter) = lower(conn.data)
+let expressUnsubscribeReplyMiddleware =
+  requireUserAndNewsletter
+    <| unsubscribeMiddleware
+    >-> writeStatus(.ok)
+    >-> respond(text: "OK")
+
+private func requireUserAndNewsletter(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, Tuple2<Database.User.Id, Database.EmailSetting.Newsletter>, Data>
+  ) -> Middleware<StatusLineOpen, ResponseEnded, MailgunForwardPayload, Data> {
+
+  return { conn in
+
+    guard let (userId, newsletter) = userIdAndNewsletter(fromUnsubscribeEmail: conn.data.recipient)
+      else {
+        return conn
+          |> writeStatus(.notAcceptable)
+          >-> respond(text: "Not acceptable")
```

Comment:
if we can't extract the user id and newsletter from the unsubscribe email, then sending back a `406` will prevent mailgun from trying again.

Reply:
Assuming headers are what it cares about we could add a `end` overload (if we don't already have it) that ends a request with headers still open, and even a `head` function that takes a status and immediately finishes the response!

``` swift
conn |> head(.notAcceptable)
```

Reply:
Ah yeah, body doesn’t matter. I could also maybe just `>-> end` It